rspace/src/main/scala/coop/rchain/rspace/history/package.scala

```diff
@@ -0,0 +1,191 @@
+  }
+
+  def lookup[T, K, V](store: ITrieStore[T, K, V], key: K)(implicit codecK: Codec[K]): Option[V] = {
+    val path = codecK.encode(key).map(_.bytes.toSeq).get
+    @tailrec
+    def loop(txn: T, depth: Int, curr: Trie[K, V]): Option[V] =
+      curr match {
+        case Node(pointerBlock) =>
+          val index: Int = JByte.toUnsignedInt(path(depth))
+          // We use an explicit match here instead of flatMapping in order to make this function
+          // tail-recursive
+          pointerBlock.toVector(index) match {
+            case None =>
+              None
+            case Some(hash: Blake2b256Hash) =>
+              store.get(txn, hash) match {
+                case Some(next) => loop(txn, depth + 1, next)
+                case None       => throw new LookupException(s"No node at $hash")
+              }
+          }
+        case Leaf(lk, lv) if key == lk =>
+          Some(lv)
+        case Leaf(_, _) =>
+          None
+      }
+    store.withTxn(store.createTxnRead()) { (txn: T) =>
+      store.get(txn, store.workingRootHash.get).flatMap { (currentRoot: Trie[K, V]) =>
+        loop(txn, 0, currentRoot)
+      }
+    }
+  }
+
+  @tailrec
+  private[rspace] def getParents[T, K, V](
+      store: ITrieStore[T, K, V],
+      txn: T,
+      path: Seq[Byte],
+      depth: Int,
+      curr: Trie[K, V],
+      acc: Seq[(Int, Node)] = Seq.empty): (Trie[K, V], Seq[(Int, Node)]) =
+    curr match {
+      case node @ Node(pointerBlock) =>
+        val index: Int = JByte.toUnsignedInt(path(depth))
+        pointerBlock.toVector(index) match {
+          case None =>
+            (curr, acc)
+          case Some(nextHash) =>
+            store.get(txn, nextHash) match {
+              case None =>
+                throw new LookupException(s"No node at $nextHash")
+              case Some(next) =>
+                getParents(store, txn, path, depth + 1, next, (index, node) +: acc)
+            }
+        }
+      case leaf =>
+        (leaf, acc)
+    }
+
+  // TODO(ht): make this more efficient
+  private[rspace] def commonPrefix[A](a: Seq[A], b: Seq[A]): Seq[A] =
+    a.zip(b).takeWhile { case (l, r) => l == r }.map(_._1)
+
+  private[rspace] def rehash[K, V](trie: Node, nodes: Seq[(Int, Node)])(
+      implicit
+      codecK: Codec[K],
+      codecV: Codec[V]): Seq[(Blake2b256Hash, Trie[K, V])] =
+    nodes.scanLeft((Trie.hash[K, V](trie), trie)) {
+      case ((lastHash, _), (offset, Node(pb))) =>
+        val node = Node(pb.updated(List((offset, Some(lastHash)))))
+        (Trie.hash[K, V](node), node)
+    }
+
+  private[rspace] def insertTries[T, K, V](
+      store: ITrieStore[T, K, V],
+      txn: T,
+      rehashedNodes: Seq[(Blake2b256Hash, Trie[K, V])]): Option[Blake2b256Hash] =
+    rehashedNodes.foldLeft(None: Option[Blake2b256Hash]) {
+      case (_, (hash, trie)) =>
+        store.put(txn, hash, trie)
+        Some(hash)
+    }
+
+  def insert[T, K, V](store: ITrieStore[T, K, V], key: K, value: V)(implicit
```

Comment:
This method is huge and as such very hard to comprehend :-(

Reply:
Code in `case Some(currentRoot) =>` could be extracted out to a function

Reply:
@dzajkowski @rabbitonweb Any ideas?

Reply:
> Code in case Some(currentRoot) => could be extracted out to a function

Snipping at the point of acquiring the current root doesn't really save us much unfortunately.

One of the things that I struggled with when pondering whether to extract smaller functions was that it would diminish my ability to see & compare how the cases are handled in context.  If I did choose to do it I would probably extract out functions for each case of the pattern match on the `tip`, but I'm still not sure...